### PR TITLE
Bug/12065 refac glob matching

### LINKED ISSUE #12065: [Refactor] Replace recursive Glob.globMatch with iterative approach
### Describe the bug

This issue is related to a recent pull request: https://github.com/opensearch-project/OpenSearch/pull/11060

Similar to core's [Regex](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java) class, there is another internal class called [Glob](https://github.com/opensearch-project/OpenSearch/blob/main/libs/common/src/main/java/org/opensearch/common/Glob.java) which also uses recursion to evaluate a glob match. This method can also be updated to use an iterative approach. 

### Additional Context

In https://github.com/opensearch-project/OpenSearch/pull/11060, there was a recursive method for evaluating core's own [Regex](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java) class. The method [simpleMatchWithNormalizedString](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java#L131-L153) was refactored from a recursive approach to an iterative approach for more predictable space and runtime complexity. 

- [ ] Refactor Glob.globMatch to use iterative method
- [ ] Update FilterPathTests to cover more scenarios

### Related component

Search

### To Reproduce

Should maintain same behavior as now

### Expected behavior

Should maintain same behavior as now

### Additional Details

_No response_

### PR DESCRIPTION
**Pull request by robinf95**
### Description
Refactored Glob.globMatch by using iterative approach

### Related Issues
Resolves #12065

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by deshsidd** [comment]
Build still failing please check and resubmit.

File: libs/common/src/main/java/org/opensearch/common/Glob.java [hunk 1]
```diff
@@ -39,47 +39,50 @@
...
-            }
-            String part = pattern.substring(1, nextIndex);
-            int partIndex = str.indexOf(part);
-            while (partIndex != -1) {
-                if (globMatch(pattern.substring(nextIndex), str.substring(partIndex + part.length()))) {
-                    return true;
-                }
-                partIndex = str.indexOf(part, partIndex + 1);
+
+        int stringIndex=0;
```
Comment (deshsidd) [inline libs/common/src/main/java/org/opensearch/common/Glob.java:57]: nit: formatting

File: libs/common/src/test/java/org/opensearch/common/GlobTests.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+
+    public void testGlobMatchWildcardAtMiddle() {
+        assertTrue(Glob.globMatch("test*ing", "testthisisaing"));
+        assertFalse(Glob.globMatch("test*ing", "testthisisa"));
+    }
+
+
+    public void testGlobMatchMultipleWildcards() {
+        assertTrue(Glob.globMatch("*test*", "thisisatesting"));
```
Comment (deshsidd) [inline libs/common/src/test/java/org/opensearch/common/GlobTests.java:41]: Is something like te**st allowed and do we test this?

Comment (robinf95) [reply libs/common/src/test/java/org/opensearch/common/GlobTests.java:41]: According to the old code, this was allowed. 
image
 I have added corresponding tests.

File: libs/common/src/main/java/org/opensearch/common/Glob.java [hunk 3]
```diff
@@ -49,37 +49,42 @@ public class Glob {
...
+            if (patternIndex < pattern.length() && str.charAt(stringIndex) == pattern.charAt(patternIndex)) {
+                stringIndex++;
+                patternIndex++;
+            } else if (patternIndex < pattern.length() && pattern.charAt(patternIndex) == '*') {
+                // wildcard found
+                wildcardIndex = patternIndex;
+                patternIndex++;
+                wildcardStringIndex = stringIndex;
+            } else if (wildcardIndex != -1) {
+                // last pattern pointer was a wildcard
```
Comment (kkmr) [inline libs/common/src/main/java/org/opensearch/common/Glob.java:73]: Can you a comment on what this block does?  My understanding is that there was not a match, so you reset pointers from last wildcard and restart the search

File: libs/common/src/test/java/org/opensearch/common/GlobTests.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    public void testGlobMatchWildcardWithEmptyString() {
+        assertTrue(Glob.globMatch("*", ""));
+        assertTrue(Glob.globMatch("a*", "a"));
+        assertFalse(Glob.globMatch("a*", ""));
+    }
+
+    public void testGlobMatchMultipleWildcardsWithMultipleCharacters() {
+        assertTrue(Glob.globMatch("a*b*c", "abc"));
+        assertTrue(Glob.globMatch("a*b*c", "axxxbxbc"));
+        assertTrue(Glob.globMatch("a*b*c", "aabc"));
```
Comment (kkmr) [inline libs/common/src/test/java/org/opensearch/common/GlobTests.java:71]: Can you add a test for match("best*test", "bestbuy")

File: libs/common/src/main/java/org/opensearch/common/Glob.java [hunk 5]
```diff
@@ -49,37 +49,42 @@ public class Glob {
...
+                patternIndex = wildcardIndex + 1;
+                wildcardStringIndex++;
```
Comment (kkmr) [inline libs/common/src/main/java/org/opensearch/common/Glob.java:75]: Should wildCardIndex be set to -1 here?  After one match, we can forever fall this if block right?

File: libs/common/src/main/java/org/opensearch/common/Glob.java [hunk 6]
```diff
@@ -49,37 +49,42 @@ public class Glob {
...
-            }
-            String part = pattern.substring(1, nextIndex);
-            int partIndex = str.indexOf(part);
-            while (partIndex != -1) {
-                if (globMatch(pattern.substring(nextIndex), str.substring(partIndex + part.length()))) {
-                    return true;
-                }
-                partIndex = str.indexOf(part, partIndex + 1);
+
+        int stringIndex = 0;
```
Comment (cwperks) [inline libs/common/src/main/java/org/opensearch/common/Glob.java:57]: Is this logic any different then [`Regex.simpleMatchWithNormalizedStrings`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java#L131-L161). If they are identical then can they be written and tested in a common place?

**Comment by kkhatua** [comment]
@cwperks 
The original issue https://github.com/opensearch-project/OpenSearch/issues/12065 seems to be resolved by https://github.com/opensearch-project/OpenSearch/pull/13104 . 
Can we close this PR as it is not needed?

**Comment by cwperks** [comment]
@kkhatua Yes, this can be closed.

**Pull request closed without merging**